### ISSUE TITLE
tf.io.gfile.walk broken on Windows

### ISSUE THREAD

**Issue by adriangb** [tensorflow/tensorflow#48086]
It seems that the path isn't split correctly. Filenames come with an extra `\` prefix. Minimal reproducible example:

```python3
import os
import tensorflow as tf


tf.io.gfile.makedirs("ram://folder")
with tf.io.gfile.GFile("ram://folder/file.txt", mode="w") as f:
    f.write("data")

for root, _, filenames in tf.io.gfile.walk("ram://folder"):
    for filename in filenames:
        assert tf.io.gfile.exists(os.path.join(root, filename))
```

This passes on *nix but not on Windows. Here is a quick CI run in GitHub actions showing this: https://github.com/adriangb/tensorflow-test/actions/runs/688190284

ccing @mihaimaruseac @bhack

**Comment by adriangb** [comment]
Some relevant discussion: https://github.com/tensorflow/tensorflow/pull/39609#discussion_r600667357

**Comment by bhack** [comment]
Thanks

**Comment by bhack** [comment]
The problem in this specific example is that you could not use `os.path.join` cause it will add in python `\` native separator on Windows that is not the ram filesystem separator `/`.
See https://github.com/bhack/tensorflow-test/blob/master/test.py

**Comment by adriangb** [comment]
Hmm good point, I was trying to simplify the example to not include model saving, but maybe that's the source of the problem? I reverted to [8c3dae](https://github.com/adriangb/tensorflow-test/blob/8c3dae59efbbf62e746b24ab2227307fad072466/test.py), which [does reproduce the issue](https://github.com/adriangb/tensorflow-test/runs/2198176548?check_suite_focus=true).

**Comment by bhack** [comment]
Yes I suppose that the issue is more in the pythoh native join in model save and load. E.g. see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/saving/saved_model/save.py#L96.

X filesystem separator in c++ is here but I don't think it is exposed in pyhton:
https://github.com/tensorflow/tensorflow/blob/343cb03f21c73cfe84e47ab2974a6faa1ad80972/tensorflow/core/platform/file_system.h#L394-L399

**Comment by adriangb** [comment]
I have a reproducible version without model saving (although of course these two could be unrelated, you make a good point above):

[Test passing on *nix](https://github.com/adriangb/tensorflow-test/runs/2198281295?check_suite_focus=true)
[Test failing on Windows](https://github.com/adriangb/tensorflow-test/runs/2198281305?check_suite_focus=true)

Source (also [here](https://github.com/adriangb/tensorflow-test/blob/02a7cdb7dbb386366573b775a8231ac64dfacd85/test.py)):

```python3
import tensorflow as tf


tf.io.gfile.makedirs("ram://test/inner")

with tf.io.gfile.GFile("ram://test/inner/file.txt", mode="w") as f:
    f.write("data")

for root, _, filenames in tf.io.gfile.walk("ram://test"):
    for filename in filenames:
        path = root + "/" + filename
        print(f"root: {root}")
        print(f"filename: {filename}")
        print(f"path: {path}")
        assert path == "ram://test/inner/file.txt"
```

**Comment by bhack** [comment]
I suspect this is another problem. 
I don't know if you can patch on the fly in you repo action  https://github.com/tensorflow/tensorflow/blob/306904197c95cc01cdcd30462fd62984329f5cef/tensorflow/python/lib/io/file_io.py#L838-L839
To print `_make_full_path` and `is_directory` result in Ubuntu/Win

**Comment by adriangb** [comment]
I should be able to. I'll give it a try tomorrow.

**Comment by amahendrakar** [comment]
> This passes on *nix but not on Windows

@adriangb,
With [TF v2.4](https://colab.research.google.com/gist/amahendrakar/244269ccabf4e7b5f175ce5eaf662264/48086.ipynb), I was able to reproduce the `AssertionError` on Linux as well. 

However, I did not face any error while running the code with the latest [TF-nightly](https://colab.research.google.com/gist/amahendrakar/6fcf1ad6081f233913593481a765c7b8/48086-tf-nightly.ipynb). Please check the linked gist for reference. 

Could you please check if you are facing the same error with TF-nightly as well? Thanks!

**Comment by bhack** [comment]
@amahendrakar We was always working to debug this with nightly. See https://github.com/adriangb/tensorflow-test/blob/master/.github/workflows/test.yml#L14

**Comment by bhack** [comment]
@adriangb Ok I've Win emulated your last example on Linux with Docker+Wine. 

You can test yourself with

```
docker run -it --rm tobix/pywine /bin/bash
wine pip install tf-nighlty-cpu
wine python <your_code_stub>.py
```
You can add debug prints to `/opt/wineprefix/drive_c/Python39/Lib/site-packages/tensorflow/python/lib/io/file_io.py` in `def _make_full_path(parent, item):`

Just for your last **specific test case** you could replace `return os.path.join(parent, item)` with `return os.path.join(parent, item).replace("\\","/")` and run your last example in the Win emulated Docker.

**Comment by bhack** [comment]
@frankchn @mihaimaruseac  As i see in your ram filesystem [tests](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/ram_file_system_test.py) do you expect an universal "/" for the ram filesystem and not an OS dependent one right?

**Comment by adriangb** [comment]
I was able to test using docker and wine, thanks for the tip @bhack !

It looks like what is happening is that this `os.path.join`:
https://github.com/tensorflow/tensorflow/blob/204082475214b3f08d1301998780592b53076951/tensorflow/python/lib/io/file_io.py#L824
Is joining `parent="ram://test"` and `item="inner"` -> `path="ram://test\inner`, which then cascades into `is_directory(path) == False`, so `\inner` gets returned as a file, etc (side note: maybe `is_directory` should check if the file exists and fail here instead of returning `False`)

It seems to me like generally the problem stems from the fact that, on windows, `os.path.join("a", "b") == "a\b"` but you are using `"/"` as the path separator for ram filesystems regardless of the platform.

**Comment by adriangb** [comment]
So it seems like 2 things are needed to solve this:
1. Some way to determine what the path separator is given the full path.
2. An implementation of `os.path.join` that accepts a path separator.

**Comment by bhack** [comment]
If you see that ram filesystem tests I mentioned above you see that it is only tested with "/" separator.

But OS dep separator it is not illegal. E.g. run in your Wine environment:

```
import tensorflow as tf


tf.io.gfile.makedirs("ram://test\\inner")

with tf.io.gfile.GFile("ram://test\\inner\\file.txt", mode="w") as f:
    f.write("data")

assert tf.io.gfile.exists("ram://test\\inner\\file.txt")

```

**Comment by adriangb** [comment]
Perhaps. But those files wouldn't exit. For example:

```python3
tf.io.gfile.makedirs("ram://test/inner")

with tf.io.gfile.GFile("ram://test/inner/file.txt", mode="w") as f:
    f.write("data")  # this is what save model does

print(tf.io.gfile.exists("ram://test\\inner\\file.txt"))
```

That is, the files are saved with `/`, but when `walk` calls `os.path.join` they end up with `\`, so they are different files.

**Comment by bhack** [comment]
I see in your old debug print with save `listing (L832) =  ['\\assets', '\\keras_metadata.pb', '\\saved_model.pb', '\\variables']`
As we told above `save` is using the native Os separator at save time right?
https://github.com/tensorflow/tensorflow/blob/94f92c9949c2dc6afe1db959e25631520602c6ea/tensorflow/python/keras/saving/saved_model/save.py#L96

**Comment by adriangb** [comment]
Maybe? It's very much possible that there is more than one issue at play here. Well, I think it's the same issue, just different places in the codebase that it's occurring. The end result is the same: the `ram://` filesystem (and presumably other non-native filesystems) are not fully functional on Windows.

**Comment by bhack** [comment]
Let's wait to understand what the original design scope was. If `ram://` works only with `/` but it isn't enforced or it need to work on native Sep by design.

**Comment by frankchn** [comment]
So the original design for the RAM file system is quite limited -- it was originally just for Cloud TPUs to have a place to write temporary files to (since Cloud TPUs don't have access to local file systems and writing to GCS is slow). 

I think if someone wants to get `ram://` working with Windows-style separators etc, we are happy to accept the controbution.

**Comment by bhack** [comment]
@adriangb As `test_savedmodel` test was passing on Win and Linux is this enough for your use case at https://github.com/tensorflow/tensorflow/pull/39609#discussion_r600667357?
https://github.com/tensorflow/tensorflow/blob/3c16284eb619732b69948f0200ee06c5dd7312d0/tensorflow/core/platform/ram_file_system_test.py#L146-L156

**Comment by adriangb** [comment]
@bhack that test is not enough for #39609.

We need to load all of `ram://my_module,` into a single contagious string of bytes so that it can be serialized. There is no other way to do this other than iterating over each file, which is exactly what `walk` is for. I don't see any way to do this without fixing `walk` or re-implementing it using `listdir`, `isdir`, etc.

@frankchn so I take it that support (or explicitly not supporting) `\` for `ram://` on Windows was not part of the original design? Does this mean that if I were to implement fixes that enabled the use case in #39609 but possibly broke other use cases involving `\` on Windows, that would be okay? Assuming it doesn't break any exisiting tests. I do not have the bandwidth to fully implement support for `ram://test\other` on Windows (or alternatively explicit lack of support), especially if that includes submitting RFCs or editing source on the C++ side. I may however be able to find a way in which we can edit 1-2 lines of Python to at least make #39609 work.

**Comment by bhack** [comment]
@adriangb Can you extend that test or a new one in the same file with a new PR just to cover your case. If I have time to explore a C++ fix I need to use the CI cause Wine is too slow to compile a large source code like TF without an available cache produced in the same environment.

**Comment by adriangb** [comment]
We have tests in #39609 . Why do we need new tests or a new PR?

Thanks for looking into a fix.

**Comment by bhack** [comment]
I meant that we need a new test (with a new PR) in `ram_filesystem_test.py` to test the isolated feature that you need in the ram_file_system.

**Comment by adriangb** [comment]
I don't think this is necessarily a feature. Depending on whether it was intentional for `\` to be supported on `ram://` or not (it sounds like it was neither), it is either a design oversight or a bug.

**Comment by adriangb** [comment]
I think the easiest solution here is to implement a TensorFlow specific `os.path.join`, maybe `tf.io.gfile.join`:

```python
import os
from posixpath import join as urljoin

def join(*paths):
    root = str(paths[0])
    if root.startswith("ram://") or root.startswith("gs://"):
        return urljoin(*paths)
    return os.path.join(*paths)
```

I'm not sure if `root.startswith("ram://")` is the best we can do here, that will be up to your team to decide.

The more difficult part of this will be replacing `os.path.join` with `tf.python.io.file_io.join` all around the codebase.

Then finally comes the issue of testing, features and backwards compatibility. I suppose this might break things for anyone relying on `ram://test\other` like behavior on Windows. But since this was never documented (or intentional) I think that should be okay. We can/should add a small test for this, but I think the test should focus on the implementation of `tf.python.io.file_io.join`, not on the behavior of other things like `walk` or `SavedModel` (i.e. let's not make any promises about those).

Does this sound reasonable?

**Comment by bhack** [comment]
I don't know if `model.save` it is supported on GCS  see https://github.com/tensorflow/tensorflow/issues/36453. 
Probably `tf.saved_model.save` is going to work cause it use c++ impl with `io::JoinPath`.
But if you still need `model.save` in your case and `model.save` is using os native join at the Python level probably it is easier to let r`ram://` to support Win native separator as `ram://` is currently not opinionated. 
Let me know if you have a test for `tensorflow/tensorflow/core/platform/ram_file_system_test.py`

**Comment by adriangb** [comment]
My use cases don't require saving anything to gcs. I only threw `root.startswith("gs://")` in there because I thought you'd want it implemented (and since it's only a couple extra characters).

I also don't care about `tf.saved_model.save` vs `model.save` vs. any other way to save things using SaveModel, they would all be equally fine for #39609. But I tested `tf.saved_model.save`  and it does not solve the problems surrounding the use of `os.path.join` on `ram://` in Windows. For example, see here:

https://github.com/tensorflow/tensorflow/blob/3c16284eb619732b69948f0200ee06c5dd7312d0/tensorflow/python/saved_model/save.py#L878

The only solution I see here is what I proposed above, implementing a version of `os.path.join` that is aware of using `/` and not `\` for the `ram://` filesystem on Windows. We can write a test for this and put it wherever you deem best (I would suggest it live with the rest of the tests that correspond to `tf.python.io.file_io` since it is not specific to `ram://`).

**Comment by bhack** [comment]
What I meant is that using  the ticket info and if a test like this will pass also on GCS I suppose that it is not using `os.join` at the python level cause GCS doesn't work with `\`. 

https://github.com/tensorflow/tensorflow/blob/3c16284eb619732b69948f0200ee06c5dd7312d0/tensorflow/core/platform/ram_file_system_test.py#L146-L156

And probably it is why `model.save` doesn't work also with GCS (I've not tested this case so it is just a reference to the mentioned ticket ticket).

Your mentioned `os.path.join` is only for the extra debug info case so probably it could now work on GCS but the default value is `save_debug_info=False`.

I think that we need to separate Walk fix vs if different save methods add os native sep in python.

My hypothesis In the case we don't have the os native separator at python level we could make a specific test and fix only the Walk case.

This is why I want to have am addition test to fail for `ram_filesystem_test.py`

**Comment by adriangb** [comment]
So you want a test that explicitly checks for support of the `ram://folder\subfolder` pattern?

**Comment by bhack** [comment]
I suppose that we could have more candidate workarounds with `ram://folder\subfolder`, as I don't see a specific `IsDirectory` in `ram_file_system.h` I suppose it rely on Posix and Window fs `isDirectory`.

**Comment by adriangb** [comment]
I don't see any way to write specific tests for that that don't explicitly suggest support for that pattern.

That is why my proposal focuses on the limiting the use of that pattern within the rest of the ecosystem. That way no decision needs to be made on explicit support of that pattern and no fixes need to be applied to existing functionality.

**Comment by adriangb** [comment]
> Your mentioned os.path.join is only for the extra debug info case

I think there may be more:

https://github.com/tensorflow/tensorflow/blob/b221f78a300a6e50564acfd94dae6b608008474d/tensorflow/python/saved_model/save.py#L1131-L1133

**Comment by bhack** [comment]
Check if this test extension is enough for your case https://github.com/tensorflow/tensorflow/pull/48124 
But it seems to me from the logs that ramfs tests are excluded on Windows CI.

**Comment by Venkat6871** [comment]
Hi **@adriangb** ,
Can you please try with recent TF version? I tried with TF2.16.1 and I cannot reproduce the error. Please check the [gist](https://colab.sandbox.google.com/gist/Venkat6871/74979dcfc754fb863ed30e7f992b5b7b/48086_2-16-1-v.ipynb) here.

Thank you!

**Comment by mihaimaruseac** [comment]
@Venkat6871 the Colab is running under linux. This issue is about Windows.
